src/simulation.hpp

```diff
@@ -2946,11 +2952,46 @@ template <typename problem_t> void AMRSimulation<problem_t>::doDiagnostics()
 				amrex::MultiFab::Copy(*diagMFVec[lev], mf, mf_idx, v, 1, 1);
 			}
 		}
+		diagMFVec_ptr = GetVecOfConstPtrs(diagMFVec);
 	}
 
+	// Prepare common diagnostic data
+	auto const geoms = Geom(0, finestLevel());
+	auto const ref_ratio = refRatio();
+
 	for (const auto &diag : m_diagnostics) {
 		if (diag->doDiag(tNew_[0], istep[0])) {
-			diag->processDiag(istep[0], tNew_[0], GetVecOfConstPtrs(diagMFVec), m_diagVars, simulationMetadata_);
+			// Set common diagnostic data (including simulation pointer)
+			diag->setDiagData(this, &diagMFVec_ptr, &m_diagVars, &geoms, &ref_ratio, &simulationMetadata_);
+
+			// Call the appropriate template processDiag for each diagnostic type
+			// All diagnostics now have a unified API: processDiag<problem_t>(nstep, time)
+			auto *plotfileDiag = dynamic_cast<DiagPlotfile *>(diag.get());
```

Comment:
This looks reasonable. However, I think this dynamic dispatch workaround is not actually necessary for DiagBase and its child classes, since (unlike the AMReX ParticleContainers) we can just mark `processDiag` as virtual in `DiagBase`.

Calling a virtual function once for the whole simulation has completely negligible performance cost. It's only an issue if you call virtual functions for every cell (i.e., inside a ParallelFor), or maybe in some very extreme cases, per FAB.